### PR TITLE
[CLD-4809] Migrate to GitHub actions

### PR DESCRIPTION
**Pull request by phoinixgrr**
#### Summary

Migrating CI from CircleCI to GitHub Actions. Part of the Unified CI initiative. 

Review by commit is recommended:

Before: 
Screenshot 2023-02-08 at 12 52 41

After:
Screenshot 2023-02-08 at 12 52 07

#### Ticket Link

Ticket: https://mattermost.atlassian.net/browse/CLD-4811

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (hmhealey) [inline .github/workflows/ci.yml:171]: Should this expire at some point? I don't know if GitHub limits us on stored artifacts or anything

File: .github/workflows/ci.yml
Comment (hmhealey) [inline .github/workflows/ci.yml:208]: I actually just changed this one in the CircleCI config since I found out some people use branches named `cloud-`
```suggestion
           github.ref_name == 'cloud' ||
```

File: .github/workflows/performance-benchmarks.yml
Comment (hmhealey) [inline .github/workflows/performance-benchmarks.yml:3]: I don't know if there's any way for us to do this differently, but with CircleCI, we actually check out the web app  and set up Cypress before waiting for the build to complete. Since we're now running this all after the build completes, this'll take slightly longer now than it did before

File: .github/workflows/performance-benchmarks.yml
Comment (hmhealey) [inline .github/workflows/performance-benchmarks.yml:36]: Could we actually upload the server bundle in the new `prepare-docker-build` job as an artifact and then redownload it here instead of having to recreate the server bundle here?

File: package.json
Comment (hmhealey) [inline package.json:234]: What's this change for?

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml
Comment (spirosoik) [reply .github/workflows/ci.yml:15]: It is a good security practice to pin the versions.

File: .github/workflows/ci.yml
Comment (toninis) [reply .github/workflows/ci.yml:6]: It is possible but we do not recommend it . We can discuss this after we adopt GHA in general and track the usage . One of our goals is also cost reduction . 
cc: @mattermost/core-build-engineers

File: .github/workflows/ci.yml
Comment (toninis) [reply .github/workflows/ci.yml:22]: Caching speeds up things in order not to download/install node modules again since it is a costly operation . This is also up for discussion

File: .github/workflows/ci.yml
Comment (toninis) [reply .github/workflows/ci.yml:171]: There is a global expiration of 90 days . But yes it should . What do you suggest ? I think 14 days is ok

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml:3]: Yes that is correct.

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml:15]: A blogpost to provide some context for this decision 
https://michaelheap.com/ensure-github-actions-pinned-sha/

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml:171]: This is the current policy. 
https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy

File: .github/workflows/performance-benchmarks.yml
Comment (phoinixgrr) [reply .github/workflows/performance-benchmarks.yml:36]: This is an effort, to replicate the functionality of CircleCI to GitHub Actions. This is currently the logic implemented of the process in CircleCI. 
If there is need to improve the logic/workflow , I'd suggest we discuss this over a second iteration.

Let me know @hmhealey

File: .github/workflows/performance-benchmarks.yml
Comment (phoinixgrr) [reply .github/workflows/performance-benchmarks.yml:3]: We could indeed run the Cypress setup, prior to artifact generation , that would save about 70 seconds. 
Screenshot 2023-02-09 at 11 11 27


But I'd suggest keeping the workflow as simple as possible, instead of constantly probing GitHub API to detect job completed. 

The gains are minimal (1 minute), and after all, the execution time of the `Jest Tests` is taking longer than that. Therefore no actual gains are presented.

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml:22]: If cache hits, then `npm install` will not run. 

https://github.com/mattermost/mattermost-webapp/pull/12182/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR30-R31

There is logic injected, that makes sures that node_modules and node_dependencies step will be skipped in the event of cache hit. 
```
if: steps.npm-cache.outputs.cache-hit != 'true'
```

Caching node modules, and node dependencies is considered as best practise, since it can save CPU cycles/network calls, thus reduce the CI execution time.

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml]: Proposed change is committed.

**Review by spirosoik — APPROVED** [review]
Minor small changes. Nothing really critical, awesome job 🚀

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml:130]: I've reviewed the code for this action. 
https://github.com/SimenB/github-actions-cpu-cores/

It's pretty safe to use, and we are already are SHA pinning it. 

I would prefer to use a public Github Action as blackbox (inputs/outputs), instead of hardcoding some shell commands to detect cores `cat proc/cpuinfo | grep | wc -l` etc.  
Moreover the ghAction It clearly provides context on what it does. 

Let me know @spirosoik . If you are strongly opinionated on this, I can manually detect with Bash.

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml:189]: Nice Catch. Yes I will change that to support Prefix.

File: .github/workflows/ci.yml
Comment (toninis) [reply .github/workflows/ci.yml:130]: We can always hard code it to 2 cores . Since this is the core count on the Free GIthub Runners

File: .github/workflows/ci.yml
Comment (toninis) [reply .github/workflows/ci.yml:189]: @phoinixgrr Did we fix this ?

File: .github/workflows/ci.yml
Comment (toninis) [inline .github/workflows/ci.yml:141]: I know we are just transferring from CIrcleCI but I would prefer to build and push to S3 after all the tests have passed . 
@hmhealey for feedback . 
If we have performance issues we can still bump the resources on our Github Runners to speed tests . 
cc: @phoinixgrr

**Review by toninis — APPROVED** [review]
🚀 Excellent work . 

I only have one consideration around building on every commit . I think we should build only after tests have passed .

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml:130]: Not strongly opinionated, But I'd prefer to keep it dynamically injected during CI Runtime. 

The current execution time in CircleCI is less than 3 mins due to larger runners used (8 cores). We execute this step in 10 minutes due to the default runners GitHub Runners used (2 cores). If there is need to execute this faster, the only change needed, would be to change the `runs_on` directive for this job, and the pipeline will take care of the rest. 

The cost for detecting cores, is negligible (less than 1 sec) 
Screenshot 2023-02-14 at 11 01 45

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml:189]: Not yet. Will soon push a commit to tackle this.

File: .github/workflows/ci.yml
Comment (spirosoik) [reply .github/workflows/ci.yml:130]: I will echo what Akis mentioned about dynamically inject it. Yeah I checked the code too, seems pretty safe.

File: .github/workflows/ci.yml
Comment (spirosoik) [reply .github/workflows/ci.yml:189]: let's fix this

Comment (phoinixgrr) [reply .github/workflows/ci.yml]: Tackled with: https://github.com/mattermost/mattermost-webapp/pull/12182/commits/b64882c068b0b73f637a7c133b5fb93ebc593631

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml:141]: I agree. 
@hmhealey Let us know your thoughts on this recommendation. 

If Harrison confirms, To keep the scope of this PR limited, Let's tackle this in a Separate PR, after this gets merged

**Review by spirosoik — APPROVED** [review]

File: .github/workflows/ci.yml
Comment (hmhealey) [reply .github/workflows/ci.yml:15]: Oh. I see. That seems kind of gross, but I get why we have to do that. I guess there's also no way to centralize those numbers so that we don't need to repeat them all over the configuration?

File: .github/workflows/ci.yml
Comment (hmhealey) [reply .github/workflows/ci.yml:6]: I'd be fine with that, but I am a bit curious about the whole thing.

Do we pay anything if its run on a fork though? I wouldn't have expected we'd be the ones paying for that

File: .github/workflows/ci.yml
Comment (hmhealey) [reply .github/workflows/ci.yml:22]: Ah, I missed the cache-hit part.

Unfortunately, this has some other issues since, on a cache hit, it won't cache the compiled subpackages which are symlinked from `node_modules` since those are generated as part of `make node_modules`. Those live in:
- `packages/client/lib`
- `packages/types/lib`
- `packages/components/dist`
and we'll be adding more of those over time as the web app is split into more subpackages.

Does GHA have any equivalent to the CircleCI workspaces I started using in https://github.com/mattermost/mattermost-webapp/pull/11648 where we don't need to specify explicit file paths but can still share copies of the working directory between CI jobs?

File: .github/workflows/ci.yml
Comment (hmhealey) [reply .github/workflows/ci.yml:171]: Yeah, 14 days would probably be fine. It could theoretically be helpful to save artifacts for longer, but since we only use these for test servers, we don't even really need them after a PR is closed

File: .github/workflows/ci.yml
Comment (hmhealey) [reply .github/workflows/ci.yml:141]: I need to double check my post in the Platform-Release Collaboration channel about all the artifacts in S3, but that should be fine. The only thing we might lose out on by doing that though is the ability to run test servers for PRs that have in-progress unit tests since that's the main thing that uses webapp-generated builds from S3

**Review by hmhealey — CHANGES_REQUESTED** [review]
Sorry about the delay. The only feedback I have that might still block this is about the caching changes since the way we're caching dependencies doesn't handle compiled subpackages (in other words, the subfolders of the `packages` directory) correctly

File: .github/workflows/performance-benchmarks.yml
Comment (phoinixgrr) [reply .github/workflows/performance-benchmarks.yml:3]: As discussed, I strongly recommend keeping the pipeline simple, instead of constantly probing GitHub API to detect job completed. 
This is recommended as best practise, keeping the pipeline simple, instead of introducing complexity for questionable gains. 

Let me know if we are aligned.

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml:22]: There is `upload-artifact` action, that can be used to pass data between jobs. 
https://docs.github.com/en/actions/migrating-to-github-actions/migrating-from-circleci-to-github-actions#persisting-data-between-jobs

But it is not recommended for the whole repository. It is designed for small quantity of artifacts, and not whole repositories. 

To unblock this, let me know if we are aligned removing the Caching part completely. And revisit at a later stage. Without node-modules caching, the pipeline finishes in about 14minutes (including benchmarks), therefore I feel that we hit the performance target even without caching.

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml:6]: I believe that the setting enforced by security, is for Forks not to run CI. 
I'l have to look on this and revert with an answer.

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml:141]: There is no additional cost involved running as is, since we manage to run the pipeline in the Free Tier GitHub Runners for Open Source Projects.  

I suggest merging this as is, and revisit when @hmhealey has some context to share. To keep the scope of this PR limited.

**Comment by phoinixgrr** [comment]
> Sorry about the delay. The only feedback I have that might still block this is about the caching changes since the way we're caching dependencies doesn't handle compiled subpackages (in other words, the subfolders of the `packages` directory) correctly

Hey @hmhealey. Do re-review this after the answers given. Thanks! 🙇 🚀

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml:171]: The previous link was about Cache. 
GitHub documentation concerning artifacts retention policy: https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts#about-workflow-artifacts 

_"By default, GitHub stores build logs and artifacts for 90 days, and this retention period can be customized."_

Will push a commit to retain artifacts for 14 days.

Comment (phoinixgrr) [reply .github/workflows/ci.yml:171]: Done with: https://github.com/mattermost/mattermost-webapp/pull/12182/commits/5bd42f916c529909e890ada0946c463c643e4b99

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml]: Caching disabled for node_modules with: https://github.com/mattermost/mattermost-webapp/commit/1cb8e2afb7b9fa69b45c605d1ea32a32310ea046

File: .github/workflows/ci.yml
Comment (phoinixgrr) [reply .github/workflows/ci.yml:6]: The current policy applied for `mattermost-webapp` concerning running CI under Forked Repos: 

Screenshot 2023-02-15 at 11 34 49

Let me know if there is any need for changes here.

File: .github/workflows/performance-benchmarks.yml
Comment (hmhealey) [reply .github/workflows/performance-benchmarks.yml:3]: Yep, this works for me

File: .github/workflows/ci.yml
Comment (hmhealey) [reply .github/workflows/ci.yml]: Sounds good. I think that's about as long as CircleCI+benchmarks took before as well

File: .github/workflows/ci.yml
Comment (hmhealey) [reply .github/workflows/ci.yml:6]: Ah, I see. I know I've seen other projects where a fork runs its own GHA, but that must've had different performance settings. It was also on tag, not on PR, so it was controlled purely by the fork owner and only affected their usage

**Review by hmhealey — APPROVED** [review]

**Pull request merged**
